### PR TITLE
[AMDGPU] Enable constant offset promotion to immediate FLAT

### PR DESCRIPTION
**Pull request by rampitec**
Currently it is only supported for FLAT Global.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp [hunk 1]
```diff
@@ -2055,10 +2055,20 @@ bool SILoadStoreOptimizer::promoteConstantOffsetToImm(
   if (!(MI.mayLoad() ^ MI.mayStore()))
     return false;
 
-  // TODO: Support flat and scratch.
-  if (AMDGPU::getGlobalSaddrOp(MI.getOpcode()) < 0)
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp:2059]: If not anything, checking existence of global SADDR form is a very obscure and indirect way of checking support for immediate offset and address size. This is actually how I discovered this piece of code, it just started to fail with the experimental patch.

File: llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp [hunk 2]
```diff
@@ -2055,10 +2055,16 @@ bool SILoadStoreOptimizer::promoteConstantOffsetToImm(
   if (!(MI.mayLoad() ^ MI.mayStore()))
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp:2055]: Just out of curiosity: why atomics are not supported?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp:2055]: Atomics have identical addressing modes, they should be handled the same way

File: llvm/test/CodeGen/AMDGPU/promote-constOffset-to-imm.mir [hunk 3]
```diff
@@ -212,3 +212,51 @@ body:             |
     %13:vreg_64 = REG_SEQUENCE %9, %subreg.sub0, %11, %subreg.sub1
     GLOBAL_STORE_DWORD %13, %0.sub1, 0, 0, implicit $exec
 ...
+---
+
+# GFX9-LABEL: name: diffoporder_add_flat_load
+# GFX9: FLAT_LOAD_DWORD %{{[0-9]+}}, 1000, 0,
+# GFX9: FLAT_LOAD_DWORD %{{[0-9]+}}, 0, 0,
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/promote-constOffset-to-imm.mir:219]: Needs a gfx8 run line to make sure the offset isn't introduced

**Review by arsenm — APPROVED** [review]
lgtm except for the untested no-offset case

**Pull request merged**